OvmfPkg/VirtioKeyboardDxe/VirtioKeyboard.c

```diff
@@ -0,0 +1,1560 @@
+  )
+{
+  VIRTIO_KBD_DEV  *Dev;
+
+  Dev = VIRTIO_KEYBOARD_FROM_THIS (This);
+  VirtioKeyboardUninit (Dev);
+  VirtioKeyboardInit (Dev);
+
+  return EFI_SUCCESS;
+}
+
+// -----------------------------------------------------------------------------
+// EFI_SIMPLE_TEXT_INPUT_PROTOCOL API
+EFI_STATUS
+EFIAPI
+VirtioKeyboardSimpleTextInputReadKeyStroke (
+  IN  EFI_SIMPLE_TEXT_INPUT_PROTOCOL  *This,
+  OUT EFI_INPUT_KEY                   *Key
+  )
+{
+  VIRTIO_KBD_DEV  *Dev;
+  EFI_TPL         OldTpl;
+
+  if (Key == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Dev = VIRTIO_KEYBOARD_FROM_THIS (This);
+
+  OldTpl = gBS->RaiseTPL (TPL_NOTIFY);
+  if (Dev->KeyReady) {
+    // Get last key from the buffer
+    *Key = Dev->LastKey;
+
+    // Mark key as consumed
+    Dev->KeyReady = FALSE;
+
+    gBS->RestoreTPL (OldTpl);
+    return EFI_SUCCESS;
+  }
+
+  gBS->RestoreTPL (OldTpl);
+
+  return EFI_NOT_READY;
+}
+
+// -----------------------------------------------------------------------------
+// Function converting VirtIO key codes to UEFI key codes
+#define MAX_KEYBOARD_CODE  255
+
+STATIC
+VOID
+EFIAPI
+VirtioKeyboardConvertKeyCode (
+  IN OUT VIRTIO_KBD_DEV  *Dev,
+  IN UINT16              Code,
+  OUT EFI_INPUT_KEY      *Key
+  )
+{
+  // Key mapping in between Linux and UEFI
+  // https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h
+  // https://dox.ipxe.org/SimpleTextIn_8h_source.html#l00048
+  // https://uefi.org/specs/UEFI/2.10/Apx_B_Console.html
+
+  static const UINT16  Map[] = {
+    [KEY_1]             = '1',  '2',  '3', '4', '5', '6', '7', '8', '9', '0',
+    [KEY_MINUS]         = '-',  '=',
+    [KEY_Q]             = 'q',  'w',  'e', 'r', 't', 'y', 'u', 'i', 'o', 'p',
+    [KEY_LEFTBRACE]     = '[',  ']',
+    [KEY_A]             = 'a',  's',  'd', 'f', 'g', 'h', 'j', 'k', 'l',
+    [KEY_SEMICOLON]     = ';',  '\'', '`',
+    [KEY_BACKSLASH]     = '\\',
+    [KEY_Z]             = 'z',  'x',  'c', 'v', 'b', 'n', 'm',
+    [KEY_COMMA]         = ',',  '.',  '/',
+    [KEY_SPACE]         = ' ',
+    [MAX_KEYBOARD_CODE] = 0x00
+  };
+
+  static const UINT16  MapShift[] = {
+    [KEY_1]             = '!', '@',  '#', '$', '%', '^', '&', '*', '(', ')',
+    [KEY_MINUS]         = '_', '+',
+    [KEY_Q]             = 'Q', 'W',  'E', 'R', 'T', 'Y', 'U', 'I', 'O', 'P',
+    [KEY_LEFTBRACE]     = '{', '}',
+    [KEY_A]             = 'A', 'S',  'D', 'F', 'G', 'H', 'J', 'K', 'L',
+    [KEY_SEMICOLON]     = ':', '\"', '~',
+    [KEY_BACKSLASH]     = '|',
+    [KEY_Z]             = 'Z', 'X',  'C', 'V', 'B', 'N', 'M',
+    [KEY_COMMA]         = '<', '>',  '?',
+    [KEY_SPACE]         = ' ',
+    [MAX_KEYBOARD_CODE] = 0x00
+  };
+
+  // Check if key code is not out of the keyboard mapping boundaries
+  if (Code >= MAX_KEYBOARD_CODE) {
+    DEBUG ((DEBUG_INFO, "%a: Key code out of range \n", __func__));
+    Key->ScanCode    = SCAN_NULL;
+    Key->UnicodeChar = CHAR_NULL;

```

Comment:
When setting ScanCode and UnicodeChar to NULL unconditionally at the start at the function you can remove a number of lines further down because you only need the non-NULL values then.

Reply:
Valid point, fixed